### PR TITLE
sql: add SHOW COMMIT TIMESTAMP to retrieve a causality token

### LINKED ISSUE #79591: sql: provide a mechanism for determining the hybrid-logical timestamp of a committed transaction
When a client is using the hybrid-logical timestamps (`crdb_internal_mvcc_timestamp`) to directly control optimistic concurrency, it is necessary to determine the timestamp that data is committed at. Consider the following client flow:

 - insert a row, caching the inserted value and the timestamp at which it was committed
 - determine that the row should be updated or deleted based on the cached data
 - attempt to update or delete the row contingent on the timestamp not having changed

Today, there is no simple `INSERT ... RETURNING crdb_internal_mvcc_timestamp`, and using `SELECT ... , crdb_internal_mvcc_timestamp` during a transaction yields the provisional timestamp, which is not safe to use. [https://github.com/cockroachdb/cockroach/issues/58046] Using `cluster_logical_timestamp()` in the transaction that issues the `INSERT` is valid, but constrains the system and likely results in lower performance.

Some other mechanism for figuring out the timestamp to make this simple for a client would be superb.

/cc @ajwerner 

Jira issue: CRDB-14933

### PR DESCRIPTION
**Pull request by ajwerner**
Fixes https://github.com/cockroachdb/cockroach/issues/79591

Relates to https://github.com/cockroachdb/cockroach/issues/7945

Release note (sql change): A new sql statement `SHOW COMMIT TIMESTAMP` has been
added. This statement can be used to retrieve the commit timestamp of the
current explicit transaction, current multi-statement implicit transaction, or
previous transaction. The statement may be used in a variety of settings to
maximize its utility in the face of connection pooling.

When used as a part of an explicit transaction, the statement implicitly
commits the transaction internally before being able to return a causality
token. This is similar to the `RELEASE cockroach_restart` behavior; after
issuing this statement, commands to the transaction will be rejected until
`COMMIT` is issued.

When used as part of a multi-statement implicit transaction, the statement
must be the final statement. If it occurs in the middle of a multi-statement
implicit transaction, it will be rejected with an error.

When sent as a stand-alone single-statement implicit transaction, it will
return the commit timestamp of the previously committed transaction. If there
was no transaction on the connection, or the previous transaction did not
commit (either rolled back or encountered an error), the command will fail with
an error code 25000 (`InvalidTransactionState`).

The `SHOW COMMIT TIMESTAMP` statement is idempotent; it can be issued
multiple times both inside of and after transactions and will return the same result.
One rough edge is that the `cockroach sql` cli command will, by default, send
statements on behalf of the user which will lead to repeated issuances of 
`SHOW COMMIT TIMESTAMP` from the CLI returning different values. If one
disables syntax checking with `\set check_syntax=false` and one changes their
prompt1 to not require a query, perhaps with `\set prompt1=%n@%M>;`, the
command will become idempotent from the CLI.

### CODE REVIEW

File: pkg/sql/conn_executor.go [hunk 1]
```diff
@@ -3206,6 +3213,52 @@ func (ex *connExecutor) runPreCommitStages(ctx context.Context) error {
...
 
+func (ex *connExecutor) handleCommitWithCausalityToken(
+	ctx context.Context, res RestrictedCommandResult,
+) (fsm.Event, fsm.EventPayload, error) {
+	res.ResetStmtType((*tree.CommitTransaction)(nil))
+	err := ex.commitSQLTransactionInternal(ctx)
+	if err == nil {
+		res.SetColumns(ctx, colinfo.ResultColumns{
+			{
+				Name:   "causality_token",
```
Comment (stevekuznetsov) [inline pkg/sql/conn_executor.go:3224]: One area where a new user to the system can get hung up is all of the different ways that these timestamps are referred to - `crdb_internal_mvcc_timestamp`, `cluster_logical_timestamp`, "hybrid-logical clock timestamps", _etc_. While this result column name is specific and correct, it does add another descriptor to this concept and require that the user know how to connect the dots.

Comment (ajwerner) [reply pkg/sql/conn_executor.go:3230]: Fair, let's see what other folks say regarding names. I feel no strong attachment towards this one.

**Comment by ajwerner** [comment]
I'm thinking I'd like to extend this to work also in the case where you issue an implicit transaction with the special statement as the last statement in the simple protocol. This should be relatively simple to detect, but it probably means pushing the logic to detect our special statement into the `pgwire` package.

**Comment by michae2** [comment]
Could it also be extended to an implicit transaction for a statement like `INSERT ... RETURNING crdb_internal.commit_with_causality_token()`? (Maybe that's outside the scope of this PR...)

This is cool!

**Comment by ajwerner** [comment]
> INSERT ... RETURNING crdb_internal.commit_with_causality_token()

I think this is hard to reason about. Instead I'm thinking of supporting a final statement that looks like the one we support in explicit transactions which can be used to regain the autocommit behavior in simple protocol multi-statement batches.

**Comment by tamird** [comment]
Would it be possible to spell this `COMMIT RETURNING`? That syntax may extend to non-transactional statements better.

/cc @jakedt

**Comment by ajwerner** [comment]
> Would it be possible to spell this `COMMIT RETURNING`?

It gets tricky when you think about the wire protocol. I don't think it's valid to send `DataRow` after a non-data returning command like `COMMIT`. One option is to treat it exactly as the current PR does, which is to say, returning that row as thouhg `COMMIT RETURNING CAUSALITY TOKEN` were syntactic sugar for the above `SELECT crdb_internal.commit_with_causality_token();` and in the wire protocol gets called a `SELECT 1` and tells the driver that it's still in a transaction. 

```
{"Type":"DataRow","Values":[{"text":"1652986747918593000.0000000000"}]}
{"Type":"CommandComplete","CommandTag":"SELECT 1"}
{"Type":"ReadyForQuery","TxStatus":"T"}
``` 

I haven't played with it, but I suspect that there are drivers out there which would be unhappy with this behavior. It's also hard to imagine how to work this into drivers which manage transaction lifecycles themselves. Is the idea that if you had a `pgx.Tx` that you'd subsequently issue a `Commit()` after you issued the above using `QueryRow`?

Like if instead it sent back the below, which is data followed by saying there was a `COMMIT` command and now the connection is not in a transaction, I can't imagine that drivers will be happy. 

```
{"Type":"DataRow","Values":[{"text":"1652986747918593000.0000000000"}]}
{"Type":"CommandComplete","CommandTag":"COMMIT"}
{"Type":"ReadyForQuery","TxStatus":"I"}
```

In summary, I guess I'm okay with different syntax but it's a little weird to treat something with a `COMMIT` as a select. I may have some things wrong or misunderstand the ask, and I'd love to be schooled on why my analysis is bunk.

**Comment by tamird** [comment]
Maybe it's useful to explain the motivation - in the current proposal where the `SELECT` and `COMMIT` are separate operations, doesn't the `SELECT` cause the transaction to be become non-retryable? I've been away from this a long time, but it would seem to me that any write-only transactions would become read-write using this (SELECT) syntax, which is undesirable in some cases.

**Comment by ajwerner** [comment]
> doesn't the SELECT cause the transaction to be become non-retryable?

The `SELECT` only causes the transaction to become non-retryable if it returns rows. This special select only returns its row if the transaction successfully commits; the select actually internally is committing the transaction. This special select doesn't cause the transaction to become read-only, it causes the transaction to be unusable in any way except `COMMIT`. This is equivalent to what we do with `RELEASE SAVEPOINT cockroach_restart`, if you're familiar. 

> it would seem to me that any write-only transactions would become read-write using this (SELECT) syntax

What is a write-only transaction in SQL? It's not a concept in cockroach as far as I know. Is there something special about that concept? Most sql write commands do reads internally. The only way to avoid a read (i.e. do a "blind write") in SQL syntax is an `UPSERT` which provides every column.

I think if we were to introduce such a concept, it wouldn't be safe to just look at the top-level kinds. Note also that top-level selects statements can also happily do writes both via CTE and, eventually, UDFs. Even more wild is that a select statement can call certain special builtins which do schema changes.

**Comment by ajwerner** [comment]
> Maybe it's useful to explain the motivation

The motivation is to be able to provide a client with the commit timestamp of a transaction which they've issued and successfully committed. A design goal is to make it such that you don't need to be particularly careful with your connection handling in order to associate the timestamp with the right connection. This constraint leads towards a design whereby the transaction is internally in cockroach committed before the wire protocol or client driver thinks it is so that as far as the client driver is concerned, the transaction is still open when the timestamp is communicated. Any approach which attempts to retrieve the timestamp after the driver issues `COMMIT` is risky; some drivers put connections back into a pool as soon as the transaction is committed. Mucking with the `COMMIT` syntax itself is also something of a non-starter. Most drivers handle it specially as part of the transaction object lifecycle. Hence the current design. We're lying to the driver about the transaction state and then allowing the user to achieve their goal and keeping the driver happy so long as this sentinel query is the final query in the transaction. 

> - in the current proposal where the SELECT and COMMIT are separate operations,

Yeah, this is true, but the COMMIT is there just to move the connection back to a state of being usable. Nothing meaningful is allowed to happen between the two (except the other special cockroach thing: `RELEASE SAVEPOINT cockroach_restart`).

**Comment by ajwerner** [comment]
I guess revisiting this, I'm far from wedded to the actual statement we use. The current this is pretty gross. I think using a `COMMIT` tag statement will be bad for client driver interactions but I'm thinking something like `SHOW COMMIT TIMESTAMP` or something might be better than the awkward hack I have in there now and would be a lot easier to detect because it'd be its own ast node in the grammar.

**Comment by michae2** [comment]
I think using `SELECT` is ok. It's similar to `SELECT CURRENT_TRANSACTION_ID()` in [SQL server](https://docs.microsoft.com/en-us/sql/t-sql/functions/current-transaction-id-transact-sql?view=sql-server-ver15).

Interestingly, it looks like there is `pg_xact_commit_timestamp` in [postgres](https://www.postgresql.org/docs/14/functions-info.html#FUNCTIONS-COMMIT-TIMESTAMP) which I think can be used after the transaction commits (IIUC).

**Comment by bdarnell** [comment]
> I think using a COMMIT tag statement will be bad for client driver interactions but I'm thinking something like SHOW COMMIT TIMESTAMP or something might be better than the awkward hack

I think having something that looks like a `SELECT` changing the connection's transaction status to committed is also going to be bad for drivers. None of the options are great for drivers.

- It's common for drivers to issue the COMMIT statement on your behalf and discard any results, so anything based on adding results to COMMIT is going to be awkward to use. For drivers that do let you call `conn.execute("COMMIT")` yourself, though, I don't think there's a fundamental obstacle to a variant like `conn.query("COMMIT RETURNING TIMESTAMP")`. We've seen this with the RETURNING clauses elsewhere - drivers don't "know" which commands return results and which ones don't, it's determined by a combination of what the server returns and which interface the caller uses.
- Hiding the commit in a `SELECT` statement is likely to confuse a driver that may still want to issue its own COMMIT/ROLLBACK at the end of its transaction scope. I think this option is strictly worse than introducing something like `COMMIT RETURNING TIMESTAMP`. 
- An after-the-fact `SELECT/SHOW commit_timestamp()` is probably the most broadly acceptable solution driver-wise; this is a fairly common pattern in other databases (widely used in the case of mysql's `select last_insert_id()`). However, because the postgres idiom for this is `INSERT RETURNING`, postgres drivers may have less support for a post-txn query than other drivers. It may be hard to reuse the connection for this purpose before it gets returned to the pool.
- Selecting the commit timestamp in between a `RELEASE SAVEPOINT cockroach_restart` and a `COMMIT` is on paper well-tolerated by all drivers, more or less matches the expected semantics of all the statements involved (for drivers which map savepoints to nested transactions), and cleanly guarantees that you're using the right connection, but in practice we know that getting people to adopt this syntax is difficult. 

My vote is for a SELECT/SHOW statement that returns the token for the last transaction committed on this transaction. It should be supported both after a plain commit for connection pooling strategies that permit this, and after RELEASE SAVEPOINT when returning connections to the pool makes this a problem.

**Comment by otan** [comment]
(removing sql-experience whilst this in debate, feel free to re-add when it's time for another look)

**Comment by ajwerner** [comment]
> Selecting the commit timestamp in between a RELEASE SAVEPOINT cockroach_restart and a COMMIT is on paper well-tolerated by all drivers, more or less matches the expected semantics of all the statements involved (for drivers which map savepoints to nested transactions), and cleanly guarantees that you're using the right connection, but in practice we know that getting people to adopt this syntax is difficult.

I'm cool with this. The major problem I have with this is that existing libraries which exist and utilize `RELEASE SAVEPOINT cockroach_restart` won't be able to adopt the causality token without a change to that library. The motivating use case was `cockroach-go`'s `crdb.ExecuteTx`.

In the current implementation, one can do something like:

```go
var hybridLogicalTimestamp apd.Decimal
if err := crdbpgx.ExecuteTx(ctx, client, pgx.TxOptions{}, func(tx pgx.Tx) error {
	if _, err := tx.Exec(
		ctx, 
		`INSERT INTO kv (key, value) VALUES ($1, $2);`,
		 uuid.New().String(), strings.Repeat("x", size),
	); err != nil {
		return err
	}
	return tx.QueryRow(
		ctx, `SELECT crdb_internal.commit_with_causality_token()`,
	).Scan(&hybridLogicalTimestamp)
}); err != nil {
	return err
}
```

One easy answer is just to allow either order. I'm happy to be convinced that it's not worth it. I think I understand the feedback and will do something that I hope will please the crowd.

**Comment by knz** [comment]
> When sent as a stand-alone single-statement implicit transaction, it will
> return the commit timestamp of the previously committed transaction.

Maybe worth mentioning in the release note that this behavior is idempotent.

**Comment by ajwerner** [comment]
> Maybe worth mentioning in the release note that this behavior is idempotent.

Done. I'll file an issue to have the CLI warn if it's going to be issuing queries unbeknownst to the user.

**Review by rafiss — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80848)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @stevekuznetsov)

___
*[`pkg/sql/conn_executor_exec.go` line 1825 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/80848#-NHM7pGS3F3KIulFHcMC:-NHM7pGb1qfToPWk4o2b:b3srjyl) ([raw file](https://github.com/cockroachdb/cockroach/blob/25d609a319df5aa3c39129109441275400c1eab9/pkg/sql/conn_executor_exec.go#L1825)):*
> ```Go
> // CommitWait state because such a savepoint had been installed when
> // the transaction was committed via SHOW COMMIT TIMESTAMP. The act of
> // checking whether the statement should be accepted also prevents any
> ```

this seems like a confusing side effect for a caller keep track of (i.e. the act of checking being a thing that toggles the flag). would it be more natural if "accepting the statement" is the thing that toggles the flag?
___
*[`pkg/sql/conn_executor_show_commit_timestamp.go` line 104 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/80848#-NHMMafW98WbK9eviCG5:-NHMMafX1GhsEWoo_nQr:b-gku7qp) ([raw file](https://github.com/cockroachdb/cockroach/blob/25d609a319df5aa3c39129109441275400c1eab9/pkg/sql/conn_executor_show_commit_timestamp.go#L104)):*
> ```Go
> }
> 
> // execShowCommitTimestampInOpenState deals with the special statement
> ```

nit: wrong function name in the comment
___
*[`pkg/sql/conn_io.go` line 143 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/80848#-NHMNGp23c-D_KfDnyXP:-NHMNGp23c-D_KfDnyXQ:b-g81na8) ([raw file](https://github.com/cockroachdb/cockroach/blob/25d609a319df5aa3c39129109441275400c1eab9/pkg/sql/conn_io.go#L143)):*
> ```Go
> 	// simple protocol Query message that contains a batch of 1 or more queries.
> 	LastInBatch bool
> 	// LastInBatch indicates that this command contains the second-to-last query
> ```

nit: variable name in the comment
___
*[`pkg/sql/sem/builtins/builtins.go` line 195 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/80848#-NHMO6BME6d9R0t029Wp:-NHMO6BME6d9R0t029Wq:b-m27nif) ([raw file](https://github.com/cockroachdb/cockroach/blob/25d609a319df5aa3c39129109441275400c1eab9/pkg/sql/sem/builtins/builtins.go#L195)):*
> ```Go
> // committed. The implementation of this function here is a shim; the function
> // is processed explicitly by the connExecutor.
> const CommitWithCausalityTokenName = "commit_with_causality_token"
> ```

is this needed?
___
*[`pkg/sql/logictest/testdata/logic_test/show_commit_timestamp` line 211 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/80848#-NHMONqPBLhS_mglcDbS:-NHMONqPBLhS_mglcDbT:b6ojxaj) ([raw file](https://github.com/cockroachdb/cockroach/blob/25d609a319df5aa3c39129109441275400c1eab9/pkg/sql/logictest/testdata/logic_test/show_commit_timestamp#L211)):*
> ```
> subtest prepare
> 
> # You cannot prepare SHOW COMMIT TIMESTAMP because it is not preparable.
> ```

could you add a test in pgwire/testdata/ and `Parse` a SHOW COMMIT TIMESTAMP statement?

**Review by ajwerner — COMMENTED** [review]
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80848#-:-NHN9bjT7kHS7X9frCie:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @stevekuznetsov)

___
*[`pkg/sql/conn_executor_exec.go` line 1825 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/80848#-NHM7pGS3F3KIulFHcMC:-NHMxjphAoeTL1JB5u1K:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/25d609a319df5aa3c39129109441275400c1eab9/pkg/sql/conn_executor_exec.go#L1825)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

this seems like a confusing side effect for a caller keep track of (i.e. the act of checking being a thing that toggles the flag). would it be more natural if "accepting the statement" is the thing that toggles the flag?
</blockquote></details>

Done.
___
*[`pkg/sql/conn_executor_show_commit_timestamp.go` line 104 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/80848#-NHMMafW98WbK9eviCG5:-NHMynA68HNhFMksS_Bw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/25d609a319df5aa3c39129109441275400c1eab9/pkg/sql/conn_executor_show_commit_timestamp.go#L104)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: wrong function name in the comment
</blockquote></details>

Done.
___
*[`pkg/sql/conn_io.go` line 143 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/80848#-NHMNGp23c-D_KfDnyXP:-NHN9SyJ9REtHwcmHrLs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/25d609a319df5aa3c39129109441275400c1eab9/pkg/sql/conn_io.go#L143)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: variable name in the comment
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/show_commit_timestamp` line 211 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/80848#-NHMONqPBLhS_mglcDbS:-NHN4Bgm28P4Co7md8nR:be73h5m) ([raw file](https://github.com/cockroachdb/cockroach/blob/25d609a319df5aa3c39129109441275400c1eab9/pkg/sql/logictest/testdata/logic_test/show_commit_timestamp#L211)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

could you add a test in pgwire/testdata/ and `Parse` a SHOW COMMIT TIMESTAMP statement?
</blockquote></details>

Done. To be clear, you can Parse a `SHOW COMMIT TIMESTAMP` statement and in that way you can prepare it using the extended protocol, you just can't prepare it using the simple protocol. The test does exercise this via the pgx code. I've added also the 
___
*[`pkg/sql/sem/builtins/builtins.go` line 195 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/80848#-NHMO6BME6d9R0t029Wp:-NHMxgdV2C59gOcQPYPf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/25d609a319df5aa3c39129109441275400c1eab9/pkg/sql/sem/builtins/builtins.go#L195)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

is this needed?
</blockquote></details>

Done.

**Comment by ajwerner** [comment]
I filed https://github.com/cockroachdb/cockroach/issues/92260 to track an edge case. The intention is not to do anything about it but to have it on the books for some future soul.

**Review by rafiss — APPROVED** [review]
lgtm! 

> Reviewed 15 of 40 files at r3, 27 of 27 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80848#-:-NHV1TUd6H7fokVhtlul:b-nzoors)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @stevekuznetsov)

**Comment by ajwerner** [comment]
TFTR!

bors r+

**Review by stevekuznetsov — APPROVED** [review]
Reviewable seemed to be upset that I had an open review - very excited to see these changes @ajwerner !

File: pkg/sql/conn_executor_commit_with_causality_token.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+	ev := eventNonRetriableErr{IsCommit: fsm.FromBool(false)}
+	payload := eventNonRetriableErrPayload{err: err}
+	return ev, payload, nil
+}
+
+// isSelectWithCausalityToken returns true if the statement is exactly the
+// following, modulo capitalization:
+//
+//   SELECT crdb_internal.commit_with_causality_token
```
[the remark this answers is not in the document]
Comment (ajwerner) [reply pkg/sql/conn_executor_commit_with_causality_token.go]: The syntax has evolved. It's now its own statement `SHOW COMMIT TIMESTAMP`.

[the remark this answers is not in the document]
Comment (ajwerner) [reply pkg/sql/conn_executor_commit_with_causality_token.go]: The syntax in the final version changed. It's now `SHOW COMMIT TIMESTAMP`.
